### ISSUE TITLE
Question: Train data one-by-one?

### ISSUE THREAD

**Issue by kevinmartinmauro** [karpathy/char-rnn#94]
Let's assume I have a thermometer in my room. I also have an Excel sheet of the last 10,000 readings from this thermometer. I get a new temperature reading every ~15 seconds.

I can train a network on the previous 10,000 values, no problem. But is it possible for me to update the weights of this network with each new temperature input? That is, rather than training a new network from scratch on 10,001 values, I want to use the already-trained network and just train one more value on it.

A related problem is this: Would it be possible for me to take this network, input the most recent 20-30 temperature values (where 20-30 is not a defined value, but varies), and see what the network predicts it will be? That is, would it be possible to write two functions that train the data one-by-one, and test the data one-by-one too? How would I go about doing this?

**Comment by hughperkins** [comment]
This is called on-line learning, and there's lots of research about it, but I dont know the specific papers.  Intuitively, if you've already run several epochs using the first 10,000 samples, you'd also want to run several epochs over the new samples, and the problem is that the new samples are going to have a disproportionately large effect on the weights plausibly.

"Playing Atari with Reinforcement Learning" involves online learning.  What they do, I think, is have a circular buffer with the last 10,000 or so data points in, and train a few epochs on this as the new data points arrive.  http://www.cs.toronto.edu/~vmnih/docs/dqn.pdf  This is termed 'experience replay', and described in http://www.aaai.org/Papers/AAAI/1991/AAAI91-122.pdf

**Comment by kevinmartinmauro** [comment]
Hmm that makes sense, thanks. Probably won't work for me right now since my dataset is huge and I don't have a capable GPU... yet...

I think I can wade through the Lua code to figure my second question out, but do you have any hints to guide me? Basically I want to write a function that loads a saved network (a "checkpoint" in this terminology), gives it a certain input (such as "Hell-"), and output what it predicts (ideally, "-o").

**Comment by hughperkins** [comment]
You can train LSTMs with any type of sequence.  Words and characters both are used, and work well.  They have slightly different characteristics.  For example, if you want to do word autocompletion, you're going to feed it a partial word.  Then maybe having character by character inputs and outputs makes this easier to implement?  Also, as a bonus, the vocabulary will be much smaller, ie around 80 or so, but on the other hand, the LSTM has to learn the way that letters are grouped together to form words.  Per-word means a larger vocabulary, but the LSTM doesnt learn how to combine letters to form words, so it seems to have advantages and disadvantages.

I'm not aware of any literature comparing these approaches, though I'm quite interested in such.

**Comment by wrapperband** [comment]
I am also interested i9n that. From previous research I can see that a small meta system that learnt how many past points gives it the best future prediction, for a state, would create a system that "learns" the characteristic of the state and can better predict when the current reading is "unusual", "automatically"

**Comment by jxchen01** [comment]
I am also trying to adapt this RNN code to solve a problem, essentially the same as the temperature problem mentioned above. I am new to torch and RNN. It would be so nice if someone can share some resources about how to implement the online learning on top of this RNN code.

**Comment by mrgloom** [comment]
Neural networks are "online learning" by default, becasue at each step you can add new batch of data(and it can be just one new sample) to update weights of network (don't know if it applicable to RNN and char-rnn).

More intersting how LSTM can be applied to reinforcement learning? I see [some papers ](http://papers.nips.cc/paper/1953-reinforcement-learning-with-long-short-term-memory.pdf)on this topic but don't know if it a good idea in practise and can it be done in char-rnn?

**Comment by HanuManuOm** [comment]
How do you test whether the model is giving you the next character by prediction or not?

**Comment by hughperkins** [comment]
@HanuManuOm 

the temperature doesnt work like the 'explore' in reinforcement, that is, it doesnt set some probability of choosing randomly.   The temperature is more like the temperature in simulated annealing: it means that all values can be chosen, but the probaiblity of choosing each value is related to the softmax output of the network.  so certain values will be chosen with higher or lower probability, depending on the output of your network.
